### DISCUSSION TITLE
the performance of Jolt.js is worse than ammo.js

### DISCUSSION THREAD

**Discussion by finscn** [jrouwe/JoltPhysics.js#16] [General]
I create 2 demos : 

http://finscn.github.io/jolt-ammo/falling-ammo.html

http://finscn.github.io/jolt-ammo/falling-jolt.html

They are use the same logic and renderer.
But ammo.js is faster than jolt.

And the result of jolt with some mistakes: 
<img width="189" alt="image" src="https://github.com/jrouwe/JoltPhysics.js/assets/288367/96a8032d-a072-4b9b-b71d-7b2b1244b0a3">
the shapes will mire down in the floor .

**Comment by finscn** [comment]
the source of  this demo: 

https://github.com/finscn/finscn.github.io/tree/master/jolt-ammo

**Reply by jrouwe** [reply]
> the shapes will mire down in the floor

The floor should not be moved in this way:

https://github.com/finscn/finscn.github.io/blob/7f9ec4742021bb541516547b6d2b42c654200200/jolt-ammo/js/phy-jolt.js#L99-L107

For the floor, you need to set BodyCreationSettings.mMotionType to Kinematic and then use BodyInterface.MoveKinematic to move it to the new place. 

Bullet is also jittering like crazy when the floor is going up, I think it needs the same fix.

**Comment by finscn** [comment]
@jrouwe Thanks for your explain.   
  I've never used 3d  physics engine and 3d game engine.   I'm a 2d indie game developer  .
Now I want to choose a high performance physics engine for my new 3D game.

the visual quality of the game `horizon` on PS4 is very high , even better then many PS5's games.

It can run on PS4, I think the performance-optimization of game engine & phyics engine must be very very good.

But in my test , Jolt is not faster than Bullet(ammo) .

Is there any mistakes in my test case ?  
Or there is some issues in the wasm version ?

**Reply by jrouwe** [reply]
When shapes sink into the floor, the collision detection will be much more expensive. I don't know if this will explain the difference, but fixing that bug is the first thing you should do.

**Reply by finscn** [reply]
I've fixed it and  the performance is better .Thanks. 
(the online demo has been updated)

Have you taken some benchmark tests for performance? 
Is there a comparison of different engines (e.g. PhysX , Cannon, Rapier3d, Ammo/Bullet ...) ?

**Reply by jrouwe** [reply]
Only for the native version (not for the WASM version): https://jrouwe.nl/jolt/JoltPhysicsMulticoreScaling.pdf (end of that doc compares Jolt with PhysX and Bullet)

**Comment by finscn** [comment]
> Only for the native version (not for the WASM version): https://jrouwe.nl/jolt/JoltPhysicsMulticoreScaling.pdf (end of that doc compares Jolt with PhysX and Bullet)

I found these native test cases take full advantage of multi-core cpu.
Could the wasm version  do that ? 
In my opinion , the wasm version  will only use one core of cpu,  when it runs in browser .

If there is only one core ,  is the performance superiority of Jolt still there ? 

(English is not my native language ,so I don't know could you understand my means.)

**Reply by jrouwe** [reply]
> If there is only one core , is the performance superiority of Jolt still there ?

I don't know, try it out.

> Could the wasm version do that ?

Yes you can, but apparently it works in all browsers but Safari, see: https://github.com/jrouwe/JoltPhysics/issues/577

**Comment by LeXXik** [comment]
Made a couple of tests for primitive shapes:

Ammo: https://playcanv.as/b/c0356526
Jolt: https://playcanv.as/b/29a71707

Ammo settings: fixed timestep 1/60, max substeps 10 (should be 1 or 2 usually, depending on delta time size).
Jolt settings: fixed timestep 1/60, substeps 1.

Edit:
Interestingly, my iPhone 12 Pro Max have both tests with similar numbers in Safari, almost no difference. The difference is more visible in Chrome on PC.

**Reply by jrouwe** [reply]
Interesting! For me they run about the same (this is not so strange given that most of the optimization effort went to parallelizing the code and in some cases this results in more work needing to be done for the single threaded case). 

I don't think I can look at the source code, but one thing to be aware of is that by default the number of solver iterations is different between Bullet (btContactSolverInfo.m_numIterations = 10) and Jolt (PhysicsSettings.mNumVelocitySteps = 10 and PhysicsSettings.mNumPositionSteps = 2). Bullet combines velocity and position steps, so what you can do is set PhysicsSettings.mNumVelocitySteps = 8 to make the total number of iterations 10 for Jolt too. Also make sure CCD is either off or on for both.

These were the settings I used in the C++ performance benchmarks:

https://github.com/jrouwe/BulletPerformanceTest/blob/0a92c3d74754c026e904446cb0894216a1bcd3cb/PerformanceTest.cpp#L200-L204

vs

https://github.com/jrouwe/JoltPhysics/blob/a5f498fc48334aa30f51c4c45dcbd53ac1080620/PerformanceTest/ConvexVsMeshScene.h#L92-L96

Another thing that may make a big difference is to spread out the bodies over a larger area so that not everything is in a single simulation island.

**Reply by LeXXik** [reply]
Ah, good points! I've updated the physics system to match the Bullet based on your examples. Jolt seems to be faster now, even without spreading the bodies for smaller islands. I've updated the link to the build now (sorry, not open sourced yet, so no sources at the moment).
